### ISSUE TITLE
Allow loading chess board from image.

### ISSUE THREAD

**Issue by therealmitchconnors** [veloce/lichobile#561]
Allow users to select an image of a chess board or diagram (from camera, file, or URL), and use it to load up the Board Editor with the pictured configuration.  This feature concept was discussed originally in #526.  This issue is for further discussion of the feature.

**Comment by therealmitchconnors** [comment]
My initial plan is to use @Elucidation 's [tensorflow_chessbot](https://github.com/Elucidation/tensorflow_chessbot) Convolutional Neural Network model as a basis for this feature.  I will convert his code from Python to Java, running against tensorflow for android, and nd4j in lieu of Tensorflow on Python and numpy.  The Java program will then need to be packaged into a Cordova Plugin.  This plan will not work for iOS, as a separate plugin back end would need to be written in XCode, and I do not have any OSX or iOS devices on which to develop and test.

**Comment by therealmitchconnors** [comment]
One problem I am running into early on: Building android tensorflow applications requires a pretty specific build environment.  Bazel and JDK 8 are required, along with a few others.  @veloce, what is your current build environment and automation like?  For this feature, we could modify your build environment to meet the prerequisites of tensorflow for android, or I could manage building the cordova plugin, and you could install the jar (not sure exactly how this works with Cordova).  Or we could come up with something else.  Thoughts?

**Comment by freefal** [comment]
I would recommend that you host a service that allows an image to be uploaded and responds with a FEN. That service could then be incorporated into lila or simply routed through lila. Trying to do the work on the phone just strikes me as the wrong approach (but I'm not an ML guy).

**Comment by Elucidation** [comment]
As a soft counter to that, among several existing technologies already on phones, face-detect on most smartphones is now done using an ML model, and that's running real-time on a live view finder in many cases. The hard part is training such a model, and that's usually done offline, the actual prediction step using a trained model is usually quite light in terms of computation, just a couple sets of matrix multiplications. 

Whether this is the appropriate solution for your app, I guess would depend on both how difficult it is getting tensorflow for android working for a good range of android versions, and the pros/cons of running a server, which can easily update a model, but is a hosted service with associated issues. It may be better initially to host a server if you are planning to run on devices that are not android such as iOS. Having it run on the phone will be much quicker though, and not reliant on an internet connection and a server.

**Comment by therealmitchconnors** [comment]
I'd like to add that I already have a trained model, courtesy of @Elucidation.  For an example of an Android app using an already trained model, see https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/android .  The trick will be integrating this with the cordova app and existing build environment...

I would also be interested in hosting something on the lichess site that allows additional training of the model, if they have the processing power...

**Comment by freefal** [comment]
As I said, I'm no ML expert. I simply suspect that your model will change
over time and sending PRs to lichobile every time you improve it a bit
seems inefficient.
If you implement it as a simple HTTP API that takes image data as input and
responds with a FEN, you get the following benefit:

1) Android and iOS can both be supported without any Android hooks. Having
single-platform features is undesirable. Most (all?) of our cordova hooks
have implementations on both Android and iOS.
2) Lila (the lichess website) could potentially allow users to directly
upload images to the website and get a board editor with the position.
3) You can improve your model at any time on the server and users will
immediately get the benefit. We write the app once for the API and should
never have to update the app again.

The reason I'm steering you away from Android is that I think you are going
to spend lots of time making it work and I actually think you're better off
not doing it directly on the phone. But with all that said, you are the one
implementing this, so feel free to do as you wish.

**Comment by veloce** [comment]
I'm not that familiar with cordova plugins and build. Cordova uses gradle and the plugin system has a mecanism to add extra dependencies. 

See this doc: https://cordova.apache.org/docs/en/latest/guide/platforms/android/plugin.html#adding-dependency-libraries

I can't help you more than this, sorry. You'll have to figure out the rest on your own.

**Comment by therealmitchconnors** [comment]
@Elucidation, perhaps you could weigh in on what environment you are running the chess bot from, and what sort of processing power you need to run and train the model...

**Comment by Elucidation** [comment]
Right now [/u/ChessFenBot](https://www.reddit.com/user/ChessFenBot/) runs as a [docker image](https://hub.docker.com/r/elucidation/tensorflow_chessbot/) on a google cloud instance, running a ubuntu variant.

I happened to already have an [n1-standard-1](https://cloud.google.com/compute/docs/machine-types) instance running already for other work. It runs several of my bots and other projects.

Over the last couple weeks the average CPU utilization for this bot has been a < 2%, memory usage hovers around 170 Mb, avg 100Kb/s network input. This could probably run from a raspberry pi.

Over the last year chessfenbot has only processed around 800 chessboard images (and discarded many more failures say 10-100x), so a very light load. I will be moving my bots over to a micro-instance at some point.

Training the model is done offline, I use my home desktop for example, I could also train on a cloud instance if I needed, the problem is simple enough that it wouldn't take more than a couple minutes to hours on a CPU-only instance at the moment. Retraining a model is simple to do offline, new checkpoint files are created such as the one [here](https://github.com/Elucidation/tensorflow_chessbot/tree/chessfenbot/saved_models). Updating a model through the store may be annoying, since they are on the order of 50Mb (un-optimized) for example.

Keep in mind that I chose CNNs for fun and ease of development instead of final accuracy, a K-nearest-neighbor (KNN) type approach for image search would probably work more efficiently in the majority of cases here, I assume this is how existing chess OCR apps work. I have a feeling KNN would have issues with edge cases where boards haven't been seen before, that a CNN will handle more gracefully. On the plus side, KNN is something you could implement in javascript.

**Comment by therealmitchconnors** [comment]
It sounds like it would be worthwhile to discuss this feature with the Lila teams before moving forward.  Is the Lila channel the best place for this discussion?  @Sam, would it be useful to upload only the gradients and other metadata, rather than the full image?  I am trying to find ways to prevent massive bandwidth usage by this feature...

**Comment by Elucidation** [comment]
Probably not, since the ML model takes in the chess tile images. You could
however find and crop the chessboard from the original image, rescale to
256x256 px and pass only that. That would make bandwidth usage much simpler.

**Comment by therealmitchconnors** [comment]
This thread has gone dead for a very long time, as I have been distracted.  I have wrapped the prediction service in grpc, which can be called from any client with relatively few dependencies (ideal for avoiding writing my own cordova plugin).  

https://github.com/therealmitchconnors/tensorflow_chessbot/tree/serve

PM me for an IP Address that is currently serving.  Next step is to get the server on a load balancer, and figure out how to call grpc from lichobile... Then I have to figure out how to hand ownership of the service over to the lila folks, and get it running on their hardware.

**Comment by veloce** [comment]
> PM me for an IP Address that is currently serving.

I have no idea what you mean by that, sorry. 

For the grpc client in lichobile, I don't know really grpc but I think you'll have to write a cordova plugin. There is actually a web implementation of grpc (https://github.com/improbable-eng/grpc-web) but you need to use both their client and server.

To take a picture of a board you have this [cordova plugin](https://github.com/apache/cordova-plugin-camera).

**Comment by therealmitchconnors** [comment]
For grpc, I was planning on using the node grpc client, rather than a cordova plugin.  I think that would mean it is accessible from all target platforms, since it's a part of node.  Does that make sense?

Regarding an IP Address, I am currently serving this grpc service from AWS, and if anyone wanted to play with it, I can provide details, but not in public, as there is no security around the service yet, and I'd rather not get a deluge of random traffic.

Thanks for the help.  I hope to have a merge request to you inside of a week.

**Comment by veloce** [comment]
> For grpc, I was planning on using the node grpc client, rather than a cordova plugin. I think that would mean it is accessible from all target platforms, since it's a part of node. Does that make sense?

No I'm afraid you can't do that.  Cordova uses node.js platform for it's command line interface tools, but a cordova application is *not* as node.js application. Lichobile, as any cordova application, is a web application running in a webview, wrapped in a cordova layer that allow to access many native APIs.

If you read [this](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md), you'll see that while grpc can work in a webapp (browser or webview), you have to use grpc web protocol, which is a bit different from the native grpc protocol.

Thus is you want to use grpc with a cordova app, you have 2 choices:
1 - use [this project](https://github.com/improbable-eng/grpc-web): it provides you with a Go package that wraps a grpc.Server to allow communication with a web client (I don't know the details), and a typescript web client that allow to to write typescript code directly integrated into lichobile
2 - write a cordova plugin that would use grpc native protocol

I'd got for 1 of course, like you said you don't want to write a cordova plugin if not necessary. It might require some change to you existing server, or not, you'll have to figure it out. I think you first need to try it in a simple example web app (just a browser page including the typescript grpc web client) to test it and make sure it works.

When you have that, then integration with lichobile can be started, with the use of the cordova camera plugin. I can help you starting from this point, if you have questions regarding lichobile code.

At this time, there will be also the question of hosting of the server code. I can't help you with that, and I don't know whether lichess can host it. I suggest you to contact directly the lichess team for that (contact@lichess.org).

**Comment by therealmitchconnors** [comment]
Thanks for clarifying.  I will see about the grpc-web stuff...

**Comment by therealmitchconnors** [comment]
I was wondering if you might provide input regarding an alternative design.  I chose to use grpc for this service because it can represent binary data in an extremely compact format, but perhaps it would be simpler to stick with the api format already in use, and encode the images into b64 json?

This would mean a 33% increase in payload size, and some performance degradation, but would allow us to implement the API in a manner consistent with everything else...

**Comment by veloce** [comment]
Indeed. And actually you can even upload an image in binary format with a regular xhr, [as explained here](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Sending_and_Receiving_Binary_Data).

This feature is relatively new in browser, but the good news is that is it fully supported on lichobile, because it runs on iOS >= 10.2 and android chrome 53.

**Comment by veloce** [comment]
Another useful link: https://cordova.apache.org/blog/2017/10/18/from-filetransfer-to-xhr2.html

I currently use the deprecated file-transfer-plugin, and have created #734 to address it.

**Comment by pociej** [comment]
Hi @veloce @therealmitchconnors. Let me ask what is status of this?

**Comment by therealmitchconnors** [comment]
@pociej it's stalled unfortunately.  I just never found the time...

**Comment by infinitless** [comment]
Not only scan position - but play a full game with the camera :) Has been done many times on a PC with a webcam. But would be absolutely awesome to have via a phone camera. Thereby making a laptop redundant and making chess truly smart and mobile at the same time.

**Comment by jamesqo** [comment]
I would absolutely love a feature like this.
